# Bias Detection in NLP: Concepts and a Simulated Audit

## 📚 Learning Objectives

By completing this notebook, you will:
- Understand the main ways bias shows up in NLP models and datasets
- Quantify association gaps on a **small simulated example** (the numbers below are illustrative, not
  measured from a real model)
- Know the standard bias **mitigation strategies** and a responsible-NLP checklist

## 🔗 Where this fits

**Builds on:** Course 06 (AIAT 116) — Unit 2 (bias, fairness and group metrics) and Course 07 — Unit 2, lesson 02, since word embeddings are where the association gaps measured here live.

**Used later in:** Course 10 (AIAT 124) — Unit 4, lesson 03, which detects and mitigates the same bias in generative models.

## ⚠️ Scope note

This example works with **simulated association scores** so the auditing logic is easy to follow. Measuring
bias in *real* word embeddings (e.g. with WEAT-style association tests) is what the unit exercise
(`01_nlp_applications_ethics_exercise.ipynb`) asks you to design.

---


## 🌍 Why this matters: Amazon built a recruiting model and had to throw it away

Amazon began building an automated résumé-screening engine in 2014 — rating candidates one to five stars,
like products. It was trained on ten years of résumés the company had received, and because the applicant
pool for technical roles had been overwhelmingly male, the model learned that being male was a feature of a
good candidate. Reuters reported (Jeffrey Dastin, 10 October 2018) that the system **penalized résumés
containing the word "women's"** — as in "women's chess club captain" — and **downgraded graduates of two
all-women's colleges**. Amazon's engineers edited the model to be neutral on those specific terms, could
not guarantee it would not find some other proxy for the same thing, and scrapped the project in 2017.

Two details matter more than the headline. First, **nobody wrote a rule about gender.** The association was
learned from co-occurrence — exactly the mechanism your Unit 2 embeddings used to discover that animal
words cluster together. Second, **patching the known proxies was not enough**, because a model that has a
reason to encode gender can reach it by another route.

**What goes wrong without this lesson.** A biased NLP system does not announce itself. It reports high
accuracy, passes code review, and ships — and its errors fall disproportionately on a group of people who
have no way to see it happening. Bias auditing is the practice of *going looking* for those errors before
they reach production, because no metric on your dashboard will volunteer them.

**Comparable cases in other domains:** Bolukbasi et al. found Google-News word2vec answering *man is to
computer programmer as woman is to homemaker*
([arXiv:1607.06520](https://arxiv.org/abs/1607.06520)); Prates et al. found Google Translate rendering
gender-neutral *"they are an engineer"* as *"he"* across 12 languages
([arXiv:1809.02208](https://arxiv.org/abs/1809.02208)); and closer to home, large language models are
measurably weaker on Saudi dialects than on Modern Standard Arabic — the Absher benchmark's 18,000+
questions across six Saudi dialect regions exists precisely to turn that gap from an anecdote into a
number, and reports "notable performance gaps, particularly in tasks requiring cultural inference or
contextual understanding" ([arXiv:2507.10216](https://arxiv.org/abs/2507.10216)).

## 📥 Inputs & 📤 Outputs

**Inputs:** a small dictionary of *simulated* gender-association scores for five professions (defined in the
code below — not measured from any real model).

**Outputs:** an association-gap table computed from those scores, plus mitigation strategies and a
responsible-NLP checklist.

---


In [1]:
# Worked example: how bias shows up in NLP systems and how practitioners audit it.
# The association scores below are SIMULATED for teaching — a real audit would
# measure them from embeddings or model outputs (see the references below).

"""
Unit 5 - Example 1: Bias Detection in NLP

This example demonstrates:
1. How bias manifests in NLP systems
2. A SIMULATED gender-bias audit (illustrative scores, not a real model)
3. Mitigation strategies
"""

print("=" * 60)
print("Example 1: Bias Detection in NLP")
print("=" * 60)

# 1. Understanding Bias in NLP
print("\n1. What is Bias in NLP?")
print("-" * 60)

bias_explanation = """
Bias in NLP can manifest as:
- Gender bias (associating certain professions with genders)
- Racial bias (stereotypical associations)
- Cultural bias (favoring certain languages/cultures)
- Socioeconomic bias (class-based associations)

"""

print(bias_explanation)

# 2. Example: Gender Bias Detection
print("\n" + "=" * 60)
print("2. Gender Bias Example (SIMULATED data)")
print("=" * 60)

# SIMULATED association scores, invented for this walkthrough.
# In a real audit these would be measured from word embeddings or model outputs.
professions = ["doctor", "nurse", "engineer", "teacher", "pilot"]
gender_associations = {
    "doctor": {"male": 0.7, "female": 0.3},
    "nurse": {"male": 0.2, "female": 0.8},
    "engineer": {"male": 0.8, "female": 0.2},
    "teacher": {"male": 0.4, "female": 0.6},
    "pilot": {"male": 0.9, "female": 0.1}
}

print("\nGender associations with professions (simulated scores):")
print("Profession | Male | Female | Bias Level")
print("-" * 50)

for profession in professions:
    male_score = gender_associations[profession]["male"]
    female_score = gender_associations[profession]["female"]
    bias_level = abs(male_score - female_score)
    bias_status = "High" if bias_level > 0.5 else "Moderate" if bias_level > 0.3 else "Low"
    
    print(f"{profession:10} | {male_score:.2f} | {female_score:.2f} | {bias_status}")

# 3. Bias Mitigation Strategies
print("\n" + "=" * 60)
print("3. Bias Mitigation Strategies")
print("=" * 60)

mitigation_strategies = [
    "Diverse training data - Ensure balanced representation", "Debiasing algorithms - Remove bias from embeddings",
    "Regular auditing - Test models for bias regularly",
    "Transparent documentation - Document known biases",
    "Fair evaluation metrics - Use fairness-aware metrics"
]

print("\nStrategies:")
for i, strategy in enumerate(mitigation_strategies, 1):
    print(f"  {i}. {strategy}")

# 4. Responsible NLP Checklist
print("\n" + "=" * 60)
print("4. Responsible NLP Checklist")
print("=" * 60)

checklist = [
    "✓ Test for bias in training data", "✓ Evaluate model on diverse test sets",
    "✓ Document known limitations",
    "✓ Provide transparency in model decisions",
    "✓ Regular bias audits",
    "✓ Include diverse perspectives in development"
]

print("\nChecklist:")
for item in checklist:
    print(f"  {item}")

print("\n" + "=" * 60)
print("Example completed successfully!")
print("=" * 60)

Example 1: Bias Detection in NLP

1. What is Bias in NLP?
------------------------------------------------------------

Bias in NLP can manifest as:
- Gender bias (associating certain professions with genders)
- Racial bias (stereotypical associations)
- Cultural bias (favoring certain languages/cultures)
- Socioeconomic bias (class-based associations)



2. Gender Bias Example (SIMULATED data)

Gender associations with professions (simulated scores):
Profession | Male | Female | Bias Level
--------------------------------------------------
doctor     | 0.70 | 0.30 | Moderate
nurse      | 0.20 | 0.80 | High
engineer   | 0.80 | 0.20 | High
teacher    | 0.40 | 0.60 | Low
pilot      | 0.90 | 0.10 | High

3. Bias Mitigation Strategies

Strategies:
  1. Diverse training data - Ensure balanced representation
  2. Debiasing algorithms - Remove bias from embeddings
  3. Regular auditing - Test models for bias regularly
  4. Transparent documentation - Document known biases
  5. Fair evaluation

## 💬 Discuss

The auditing logic you just ran is simple. The judgement calls it hides are not.

1. This notebook labels a gap above 0.5 "High" and below 0.3 "Low". Where did those thresholds come from?
   Amazon's model was abandoned not because a number crossed a line but because of what the number meant
   for applicants. Propose a threshold for a Saudi employer's CV-screening tool — and justify it out loud
   to someone who will not get the interview.
2. `teacher` came out "Low" (0.40 / 0.60) and `pilot` "High" (0.90 / 0.10). Suppose the real workforce
   statistics matched those numbers exactly. Is a model that faithfully reproduces reality biased? Prates
   et al. found Google Translate's male skew *exceeded* what US labour statistics predicted — does that
   distinction change your answer, and what would you do differently in each of the two cases?
3. The mitigation list printed above recommends "debiasing algorithms — remove bias from embeddings". Read
   the limits section below before answering: if debiasing mostly *hides* the association rather than
   removing it, is it still worth doing? What would you tell a client who asks you to "debias the model" as
   a one-line deliverable?

---

## ✅ Summary

**What this notebook did:** walked through the *logic* of a bias audit — score associations, compute the
gap, classify the severity — on **simulated** numbers, and collected the standard mitigation strategies and
a responsible-NLP checklist.

**What it did NOT do:** measure bias in a real model or dataset. That requires real embeddings or model
outputs and an association test such as WEAT — designing exactly that audit is the unit exercise.

**Next steps:**
- Complete `exercises/01_nlp_applications_ethics_exercise.ipynb`
- Recall the Unit 2 embeddings notebook: embeddings learn from co-occurrence, which is precisely how they
  absorb the biases of their training text


## ⚠️ Where this breaks

**This notebook's numbers are invented.** Five professions and ten hand-typed scores, chosen to make the
arithmetic legible. A table like this can demonstrate *auditing logic*; it can never be evidence that any
model is biased. Presenting simulated scores as an audit result would itself be an integrity failure — a
real audit measures associations from actual embeddings or actual model outputs (WEAT; Caliskan, Bryson &
Narayanan, *Science* 2017), on a documented word list, with a significance test.

**"Debiasing" mostly hides the bias.** The mitigation list above recommends removing bias from embeddings.
Gonen & Goldberg tested that claim directly in *Lipstick on a Pig*
([arXiv:1903.03862](https://arxiv.org/abs/1903.03862)): after standard debiasing, "the gender bias
information is still reflected in the distances between 'gender-neutralized' words … and can be recovered
from them", so "the actual effect is mostly hiding the bias, not removing it". A model that still encodes
the association will still act on it, while passing the very test you used to declare it fixed. That is
worse than not debiasing at all, because it buys false confidence.

**A gap score is not a harm.** Blodgett et al.'s critical survey of "bias" in NLP
([arXiv:2005.14050](https://arxiv.org/abs/2005.14050)) found that the field routinely reports association
metrics without stating who is harmed, how, and why that harm is wrong. Before reporting a number, write
the sentence: *"in this deployment, this system will do X to these people."* If you cannot write it, you
have measured a curiosity, not a risk.

**Where the framework itself does not transfer.** Almost all published bias measurement is English,
binary-gendered and Western. Saudi and Gulf deployments raise categories no standard word list covers:
dialect and region, tribal and family-name signals, nationality of expatriate workers, MSA-versus-
colloquial register. There is no off-the-shelf test for these. Arabic benchmark work is starting to close
the measurement gap — a survey of Arabic benchmarks and their blind spots
([arXiv:2510.13430](https://arxiv.org/abs/2510.13430)) and QIMMA's argument for validating the benchmark
before ranking models with it ([arXiv:2604.03395](https://arxiv.org/abs/2604.03395)) — but designing a
locally meaningful audit is still engineering work, not a library call.

**And one 2026 risk this checklist predates.** A text pipeline that reads untrusted documents can be
*steered* by them: indirect prompt injection, where instructions are hidden inside the content a model is
asked to process, is now the dominant vulnerability class for deployed language systems
([arXiv:2302.12173](https://arxiv.org/abs/2302.12173)). Responsible NLP in 2026 means auditing for
manipulation as well as for bias.

## 📚 References

The papers that established bias measurement in NLP — and the critical view of the field:

1. Bolukbasi, T., Chang, K.-W., Zou, J., Saligrama, V., & Kalai, A. (2016). *Man is to Computer Programmer as Woman is to Homemaker? Debiasing Word Embeddings*. NeurIPS 2016. [arXiv:1607.06520](https://arxiv.org/abs/1607.06520)
2. Caliskan, A., Bryson, J. J., & Narayanan, A. (2017). *Semantics Derived Automatically from Language Corpora Contain Human-Like Biases*. Science, 356(6334), 183-186.
3. Blodgett, S. L., Barocas, S., Daume III, H., & Wallach, H. (2020). *Language (Technology) is Power: A Critical Survey of "Bias" in NLP*. ACL 2020. [arXiv:2005.14050](https://arxiv.org/abs/2005.14050)
4. Bender, E. M., Gebru, T., McMillan-Major, A., & Mitchell, M. (2021). *On the Dangers of Stochastic Parrots: Can Language Models Be Too Big?* ACM FAccT 2021.
